### DISCUSSION TITLE
Implement Auto Click Scriptlet

### DISCUSSION THREAD

**Discussion by BlazeFTL** [uBlockOrigin/uBlock-issues#2745] [General] [answered]
URL Shorteners Are EveryWhere Now..So Wouldnt It Be Great To Implement A Auto Click Scriptlet That Clicks Elements/Buttons Automatically?

Ex:`https://shrs.link/RGDtVf` They Have Upto 4-6 Pages Of Ads With Same Timer+Buttons.. Similarly Others Shorteners Have This Too..AG Has A Scriptlet Too But I Think It Was For Cookie Notice

So If Something Like This Is Possible
Then We Could Use `Example.com##+js(Auto-Click, #getlink, #gotolink, #others)

**Comment by iam-py-test** [comment, marked as answer]
I'm sorry, but this has already been requested: https://github.com/uBlockOrigin/uBlock-issues/issues/2229

**Comment by BlazeFTL** [comment]
Sad..I Thought Since We Have Trusted Filters There Would Be No Issue...However If Someone Has A Scriptlet For UserResourceLocation Then That Would Be Nice

**Reply by Yuki2718** [reply]
Why sad? It's still open and nobody said it will never be added.

**Comment by BlazeFTL** [comment]
@gorhill trusted-click-element Its Detected As A Popup To Both FireFox And Kiwi...Not A Normal Click ?
A Prompt Saying Popup Blocked Or Allow Popup To Open New Tab

**Reply by BlazeFTL** [reply]
[Screenshot_2023-10-19-14-40-00-814_org mozilla fenix](https://github.com/uBlockOrigin/uBlock-issues/assets/139938337/3d6257dc-bb9b-4903-bd5d-2bb4632b10d1)
On FF...In First Page Click Doesn't Work Maybe Bcz FF Thinks Its Popup

**Reply by ghost** [reply]
Probably a check for https://developer.mozilla.org/en-US/docs/Web/API/Event/isTrusted

Not user-initiated - not trusted.

**Reply by BlazeFTL** [reply]
Nothing To Do From Ubo Side ?

Disabling This Works `dom.disable_open_during_load`

**Reply by MasterKia** [reply]
@BlazeFTL Which site? We can't guess.

**Reply by BlazeFTL** [reply]
@MasterKia Check Post `https://www.reddit.com/r/uBlockOrigin/comments/17beb7c/trusted_click_elem_help/`

**Reply by BlazeFTL** [reply]
@MasterKia Am I Making Any Here ?
`https://publicearn.com/sbA5oJXw`
```
starxinvestor.com##+js('trusted-click-element', '#tp98, #btn6', '', '5000')
starxinvestor.com###tp98, #btn6:style(display: block !important;)
starxinvestor.com##.BR-Footer-Ads-content, #BR-Footer-Ads, center > h3, #skull, #content:remove()
```
In Second Step It Doesn't Auto Click
If I Use `trusted-click-element', '#btn6, #tp98', '', '5000')`
Then First Page Doesn't Auto Click
Also It Doesn't Seem To Consider The 5 Sec Delay I Set It Click Immediately

**Reply by mapx-** [reply]
for uBO you can write it as
`starxinvestor.com##+js(trusted-click-element, '#tp98, #btn6', , 5000)`

**Reply by BlazeFTL** [reply]
Yes I Did Try That...But In Second Step Click Doesn't Work

**Reply by ghost** [reply]
Looks like there are two pages, one has #tp98 the other #btn6. If you write it in one filter, on first page it will click the first selector and page reloads, so filter cannot click on second selector. On second page after reload there is no first selector so filter times out. If I try separating selectors to two filters, second page reloads after click and returns to first one  - maybe server-side timer?

**Reply by BlazeFTL** [reply]
Yeah A Delay Is Required For 'tp98'
```
starxinvestor.com##+js('trusted-click-element', '#btn6', , 1000)
starxinvestor.com##+js('trusted-click-element', '#tp98', , 4000)
```

Main Problem Is This
If The Website Takes 5 Sec To Load Then Filter Gets Discarded..So A Separate Scriptlet Would Be Great
trusted-click-element-after-load

**Comment by BlazeFTL** [comment]
@gwarser Bro Can Something Be Done About This ?
The Timer Doesn't Consider If Website Loaded Fully

`starxinvestor.com##+js(trusted-click-element, '#btn6, #tp98', , 4000)`

If The Website Takes 5 Sec To Load Then Filter Gets Discarded..So A Separate Scriptlet Would Be Great
trusted-click-element-after-load🌚

**Reply by ghost** [reply]
It's trying to click as soon as possible - watches for DOM mutations. The total timeout is 10s.

**Reply by gorhill** [reply]
It's really not clear what AdGuard's documentation says about what happens internally. It's completely not clear whether the first element is delayed or not, so I changed again to have the delay apply to the first element, because of <https://github.com/AdguardTeam/Scriptlets/blob/2a3e4456bc34eb53fbaad8075294f98780550b0f/src/scriptlets/trusted-click-element.js#L311>.

Personally, I think the scriptlet should start to process at `DOMContentLoaded` at the earliest.

**Reply by gorhill** [reply]
@ameshkov Can you clarify whether the delay applies to the first element in the list? The way I understand the implementation at <https://github.com/AdguardTeam/Scriptlets/blob/2a3e4456bc34eb53fbaad8075294f98780550b0f/src/scriptlets/trusted-click-element.js>, is that if the mutation observer finds all the elements, the first one will be click regardless of the delay. However, if the clicking is not started by the mutation observer, the first element will be delayed as per `delay`. Is it really the wanted behavior?

**Reply by stanislav-atr** [reply]
@gorhill
The scriptlet will start looking for elements asap and collect found elements.
- If delay is **not** set, the elements will be clicked on the go as they are found.
- If delay is set, the scriptlet will wait for the delay to pass before starting to click the (collected)  elements in a given sequence.
There is also a static delay between clicks of 150ms, to allow _next_ elements to load or render.

**Reply by gorhill** [reply]
@stanislav-atr Ok, I see `canClick` is used to decide whether an element can be clicked, I didn't realize this. So this clarify that the delay does apply to the first element.

**Reply by garry-ut99** [reply]
> Bro Can Something Be Done About This ? The Timer Doesn't Consider If Website Loaded Fully
> 
> `starxinvestor.com##+js(trusted-click-element, '#btn6, #tp98', , 4000)`
> 
> If The Website Takes 5 Sec To Load Then Filter Gets Discarded..So A Separate Scriptlet Would Be Great trusted-click-element-after-load🌚

Another case : 
- https://github.com/uBlockOrigin/uAssets/issues/21437
